### DISCUSSION TITLE
Simpler Firebase Deployment Strategy

### DISCUSSION THREAD

**Discussion by ih84ds** [0x80/isolate-package#117] [General]
First of all, THANK YOU SO MUCH for making this package! I was pulling my hair out trying to find an elegant solution for not being able to maintain package isolation in my yarn workspaces monorepo.

I believe I have a simpler approach for deploying firebase functions than the one that is laid out [here](https://github.com/0x80/isolate-package/blob/main/docs/firebase.md#firebase). This strategy allows you to preserve "live code updates" and allows `firebase.json` to remain unmodified.

Instead of setting `"source": "./isolate"`, you can do deployment using a script like this in the `package.json` of your workspace package:

```json
  "scripts": {
    "deploy": "isolate && firebase --config isolate/firebase.json deploy",
  }
```
This takes advantage of `firebase.json` being copied into `./isolate`. That way, it will deploy from the `source` relative to *that* `firebase.json`. Meanwhile, local dev emulators remain unaffected by the isolate stuff. 🙌 

I appreciate the fact that `firebase-tools-with-isolate` exists, but I don't want my project(s) to be dependent on a fork that may or may not have all of the latest upstream changes merged.

**Comment by 0x80** [comment]
Thanks, that's a great idea! I'll look into it. 

Of course I would also prefer not having to maintain `firebase-tools-with-isolate` 😄

**Comment by 0x80** [comment]
@ih84ds I could use a bit more info on how you managed to deploy it.

I tried to apply your strategy, but I get the following error when I try to deploy:

> Error: Failed to find location of Firebase Functions SDK. Please file a bug on Github (https://github.com/firebase/firebase-tools/).

The problem seems to be that when executing the firebase deploy command, it tries to find the firebase-functions binary in the node_modules folder relative to the "source" setting. See code [here](https://github.com/firebase/firebase-tools/blob/2e84e941339b33f4ab512d46161a4383e29a6222/src/deploy/functions/runtimes/node/index.ts#L175)

So in my case it would be looking for node_modules inside of the isolate directory.

**Reply by ih84ds** [reply]
Interesting... I have a monorepo using nx, yarn workspaces, and docker, but let me see if I can boil down the deployment process...

* `yarn install` in the root of the monorepo
* `cd packages/backend && yarn deploy --project "${PROJECT_ID}" --force`
  * where the deploy script is: `isolate && firebase --config isolate/firebase.json deploy`

My `firebase.json` contains this (among other things:

```
{
  "functions": {
    "source": ".",
    "ignore": [
      ".git",
      ".runtimeconfig.json",
      "firebase-debug.log",
      "firebase-debug.*.log",
      "node_modules",
      ".yarn/cache",
      "data",
      "docker",
      "util"
    ]
  }
}
```

**Reply by 0x80** [reply]
Hmm yeah that looks very much like that I've been trying. I'll give it another go soon.

In the meantime, I have published version isolate-package@1.23.0-0 which you can install with @next. It should pick up the firebase.json from your target package and copy it to the isolate output as part of the process. 

If you feel like it, give it a try and let me know if it works for you...

**Reply by ih84ds** [reply]
Not sure what the changes were supposed to do, but simply upgrading from `1.2.0` to `1.23.0-0` breaks my deployment.

```
4.310 info Using isolate-package version 1.23.0-0
4.314 debug Found tsconfig at: /home/node/app/packages/backend/tsconfig.json
4.360 Error: Failed to find outDir in tsconfig. If you are executing isolate from the root of a monorepo you should specify the buildDirName in isolate.config.json.
4.360     at getBuildOutputDir (file:///home/node/app/node_modules/isolate-package/src/lib/output/get-build-output-dir.ts:28:13)
4.360     at isolate (file:///home/node/app/node_modules/isolate-package/src/isolate.ts:73:32)
4.360     at run (file:///home/node/app/node_modules/isolate-package/src/isolate-bin.ts:9:3)
```

**Reply by ih84ds** [reply]
...but it also fails with `1.22.0` 😅 

```
3.884 info Using isolate-package version 1.22.0
3.889 debug Found tsconfig at: /home/node/app/packages/backend/tsconfig.json
3.930 Error: Failed to find outDir in tsconfig. If you are executing isolate from the root of a monorepo you should specify the buildDirName in isolate.config.json.
3.930     at getBuildOutputDir (file:///home/node/app/node_modules/isolate-package/src/lib/output/get-build-output-dir.ts:28:13)
3.930     at isolate (file:///home/node/app/node_modules/isolate-package/src/isolate.ts:72:32)
3.930     at run (file:///home/node/app/node_modules/isolate-package/src/isolate-bin.ts:9:3)
```

**Reply by 0x80** [reply]
Whoops. Did you mean upgrading from 1.20.0 to 1.23.0-0? Because 1.2 would be pretty old...

1.22 was supposed to handle tsconfig with extends which would prevent the outDir error if outDir was defined by the extended config.

Anyway I'll look into it

**Reply by 0x80** [reply]
Odd, I'm not sure what's up. My own projects seem to have no problem deploying with firebase-tools-with-isolate@13.29.1 (which uses isolate-package@1.22.0) or isolate-package@1.22.0 separately.

**Reply by ih84ds** [reply]
> Whoops. Did you mean upgrading from 1.20.0 to 1.23.0-0? Because 1.2 would be pretty old...

D'oh. Yep, 1.20.0!

> 1.22 was supposed to handle tsconfig with extends which would prevent the outDir error if outDir was defined by the extended config.

I'm using extends without issue in 1.20.0 🤔 

> My own projects seem to have no problem deploying with firebase-tools-with-isolate@13.29.1 (which uses isolate-package@1.22.0) or isolate-package@1.22.0 separately.

I'm not using `firebase-tools-with-isolate` (which was the reason I opened this discussion -- to point on that it works great without it).

Here is my `tsconfig.json` in my backend package, FWIW:

```
{
  "extends": "../../tsconfig.base.json",
  "files": [],
  "include": [],
  "references": [
    {
      "path": "../entities"
    },
    {
      "path": "./tsconfig.lib.json"
    }
  ]
}
```

**Reply by ih84ds** [reply]
Ohh, actually, I forgot that I configured `tsconfigPath: "tsconfig.lib.json"`. Here is that one:

```
{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "baseUrl": ".",
    "rootDir": "src",
    "outDir": "lib",
    "tsBuildInfoFile": "lib/tsconfig.lib.tsbuildinfo",
    "emitDeclarationOnly": false,
    "types": ["node"]
  },
  "include": ["src/**/*.ts"],
  "references": [
    {
      "path": "../entities/tsconfig.lib.json"
    }
  ],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts", "src/**/*.test.ts"]
}
```

**Reply by ih84ds** [reply]
I got it working (on `1.23.0` today) with `buildDirName: "lib"` instead of setting `tsconfigPath`. I had overlooked that config option when initially setting things up. I actually like that better anyway. Simpler/less coupling. Sorry for the confusion! 🙌 

If you want me to test anything else with my setup, just let me know.

**Reply by 0x80** [reply]
> I'm using extends without issue in 1.20.0 🤔

Extends was only problematic if the base config was the one defining the outDir, and your extended config didn't have it, because then isolate-package would think outDir was undefined.

With the use of `get-tsconfig` all properties should now resolve correctly.

**Reply by ih84ds** [reply]
> With the use of get-tsconfig all properties should now resolve correctly.

That doesn't seem to be the case. It stopped working after that update for my tsconfig.lib.json above. But, I'm happier with the `buildDirName` solution anyway.

**Comment by 0x80** [comment]
@ih84ds Today I managed to solve a long standing pain point. Now we are able to adhere to the officially recommended Firebase setup for monorepos. See the [announcement here](https://github.com/firebase/firebase-tools/issues/653#issuecomment-2585895455) for more info.

I actually think this is the preferred approach, because it simplifies the configuration and most importantly allows you to run multiple firebase packages together in a single emulator setup. I don't think you can achieve this without the firebase-tools fork.

But likely I will still add support for your idea of deploying directly with isolate-package by copying the local firebase.json into the isolate output, because I think I already have the logic implemented in a branch...

**Reply by ih84ds** [reply]
> But likely I will still add support for your idea of deploying directly with isolate-package by copying the local firebase.json into the isolate output, because I think I already have the logic implemented in a branch...

I'm confused because it's already doing this for me as-is.

> I actually think this is the preferred approach, because it simplifies the configuration and most importantly allows you to run multiple firebase packages together in a single emulator setup. I don't think you can achieve this without the firebase-tools fork.

Ah, interesting! I hadn't run into that issue. I'm only running the emulators from from one single backend package at the moment. But I could see how that would be problematic for multiple backends that use firestore/pubsub/etc. Nice!

**Reply by 0x80** [reply]
> I'm confused because it's already doing this for me as-is.

I must have misunderstood you then. I thought you were copying the `firebase.json` config into the isolate directory yourself. So my plan was to have isolate-package handle the step of copying the config file to the isolate dir. 

Also, if the `firebase.json` is instead located at the root of the monorepo, I wanted to make it so that it extracts the configuration for that particular functions package.

**Reply by 0x80** [reply]
> I don't think you can achieve this without the firebase-tools fork.

On second thought, I think this is probably not true.  With the unified config in the root, I think it should be possible to use the normal firebase-tools for running the emulators, and then use your deployment strategy with isolate-package for deploying the individual packages.

Only you then need to remember to not use `firebase deploy` on the root, because it won't work. So I think for consistency and simplicity I would still prefer to use the fork.

BTW the fork doesn't really change the runtime workings of firebase-tools. It takes me about 5mins to sync it to upstream whenever they publish a new version.  

It's unfortunate that NPM is not fully supported yet (lockfile versions do not match) otherwise I think they could have maybe considered it as an official solution.

**Reply by ih84ds** [reply]
> I must have misunderstood you then. I thought you were copying the `firebase.json` config into the isolate directory yourself. So my plan was to have isolate-package handle the step of copying the config file to the isolate dir.

Ah, nope. The `isolate` script was/is copying everything (including firebase.json) into the `isolate` dir already.

> Also, if the `firebase.json` is instead located at the root of the monorepo, I wanted to make it so that it extracts the configuration for that particular functions package.

I'm having trouble understanding how this could work in the general case. I guess you'd be constrained to only having deployable functions in one of the packages, and you'd have to do some fancy path translation to get the isolated deployment to work.

**Reply by 0x80** [reply]
> Ah, nope. The isolate script was/is copying everything (including firebase.json) into the isolate dir already.

Ah yes you can of course just add it to the `files` property in your manifest, and then it'll get picked up by the isolate process.

> I'm having trouble understanding how this could work in the general case.

In the root configuration you'd have "functions" as an array of configurations. Each has its "source" pointing to the deployable packages. 

So when isolate runs, is just looks up the corresponding functions configuration from firebase.json based on the target package folder name, and inject the result to a local firebase.json file and overwrite "source" to be "."

See this for more info https://firebase.google.com/docs/functions/organize-functions?gen=2nd#managing_multiple_source_packages_monorepo

**Comment by ih84ds** [comment]
Following up after I recently upgraded from v1.23.0 to v1.26.1...

I had to add this to the `package.json` of my workspace package due to a requirement by `isolate-package` introduced in v1.24.0:

```
  "files": [
    ".firebaserc",
    "firebase.json",
    "firestore.indexes.json",
    "firestore.rules",
    "lib",
    "storage.rules"
  ],
```

After that, my setup is still working great!

**Reply by 0x80** [reply]
@ih84ds Thanks for sharing! I'll try to come back to this soon, because it would be great if the fork is not needed and there are no real downsides to your approach...

If you feel like it... it would be greatly appreciated if you made PR for https://github.com/0x80/typescript-monorepo

**Reply by ih84ds** [reply]
I would like to, but right now I'm pretty swamped. Will make a note to come back and check this out when things cool down a bit. 😅

**Reply by 0x80** [reply]
Ok, no worries. I'm very familiar with the swamped feeling 😄  I'll let you know when I start working on it myself...

**Reply by 0x80** [reply]
@ih84ds I just gave it a try, but I seem to be running into two issues:

1.  The only scenario in which I see this working is if you have a single package that you'd like to deploy to firebase. In larger projects I'd like to package services per domain, in which you'd want to have your firebase files in the root of the monorepo, rather than in the individual packages, because that would be problematic for things like indexes and storage rules. Also to run everything together in the emulators I think you need to have all config combined in a single set of firebase configs.

2. I have one project in which I do deploy everything from a single package, but the monorepo is using turbo, and with the config shown below, Firebase tools complains that it cannot find the turbo.json + I need to install turbo in that package as opposed to only in the monorepo root.

```json
{
  "functions": {
    "source": ".",
    "runtime": "nodejs22",
    "predeploy": ["turbo build"]
  }
}
```

**Reply by ih84ds** [reply]
Ah, yes, I honestly haven't used firebase functions based backends for more complex projects that span multiple packages on the backend. For larger projects with many backend services/packages, I don't use firebase functions (I head straight to containerized services, run them on Cloud Run or GKE, and manage with terraform).

For your turbo project, I think the solution would be to do the build first and then isolate (and remove the `predeploy`). That way you shouldn't need turbo in isolated package in order to do the deployment.

So maybe the "solution" here would be to add a section in the docs for a "simplified" configuration that works when you only have one package that uses firebase functions?
